### PR TITLE
Database Architecture section 4

### PR DESCRIPTION
**Pull request by shubh-ranade**

### CODE REVIEW

**Review by twitu — CHANGES_REQUESTED** [review]
Requires a separate note explaining jive join before getting accepted.

**Comment by shubh-ranade** [comment]
Jive join isn't a part of Database Architecture section 4.

**Comment by twitu** [comment]
It was covered in Week 4 as part of Columnar storage. Since you wanted to understand it yourself once, after explanation. You are in the best position to record an explanation.

**Comment by shubh-ranade** [comment]
As per our discussion, we're doing seperate PRs for pending notes

File: week4/db-architecture-4.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+   - Subquery flattening, heuristic rewrites: rewrite subqueries in a optimizer friendly form (query normalization)
+
+### Query Optimizer
+
+Query optimizer's job is to transform an internal query representation into an efficient query plan which is basically data flowing through a graph of query operators. Query plans are generally compiled into some interpretable data structure (lightweight object or bytecode for example) which enables cross-platform compatibility.
+
+Optimizer's functionality can be extended to:
+1. Select query plan space e.g. left deep plans only
+2. Estimate selectivity
+3. Decide on algorithms for searching the plan space
```
Comment (twitu) [inline week4/db-architecture-4.md:36]: A small definition of plan space is needed.

File: week4/db-architecture-4.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+### Query Executor
+   - Query executor on interpretable plans works on the iterator model (volcano-style). All relational operators are implemented as subclasses of the iterator class.
+
+#### Iterators
+   - Iterators couple dataflow with control flow i.e. a tuple is returned to a parent in the graph at the same time when control is returned. Which means a single DBMS thread is sufficient for executing the entire query graph.
+   - Single threaded iterators can be extended to support parallel query execution.
+
+#### Where is the data stored?
+   - In flight tuples are stored in Buffer Pool pages (BP tuples) or copied into memory heap (M tuples)
```
Comment (neelkshah) [inline week4/db-architecture-4.md]: Buffer pools are shared across processes, heap is memory specific

Comment (twitu) [inline week4/db-architecture-4.md]: Difference between BP and memory heap is needed here.

File: week4/db-architecture-4.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+#### Handling DML queries
+
+   - SQL semantics dictate that a single SQL statement should not be able to see its own updates, careful implementation is needed to ensure this.
+   - ***The Halloween problem*** is an example of what can go wrong. The problem arises on statements like 
+      >give everyone whose salary is under $20K a 10% raise.” 
+   
+   - If a plan uses an index scan iterator over salary field piplined into an update iterator, it results in all low-paid employees receiving repeated raises until they earn more than $20K.
+   - One way to avoid this problem is not using indexes on the columns being updated, but this can be quite inefficient.
+   - Another solution is to store IDs of all tuples qualifying for updates into a temporary file and scanning the file to fetch tuples to be fed to the update operator. This works well when only a few tuples are to be updated because, in that case, the temporary file is small enough to fit entirely in the buffer pool.
```
Comment (twitu) [inline week4/db-architecture-4.md]: Solution to Halloween problem is not clear. It can be added to QA file and docs for a better answer later.

**Review by neelkshah — CHANGES_REQUESTED** [review]
changes to `index.md`
jive join

File: week4/db-architecture-4.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+- The API that access methods provide is the iterator API, with the `init()` method extended to accept a ***search argument (SARG)***.
+- There are two reasons for passing SARGs into the access method layer:
+   1. Accessing indexes like B+-trees require SARGs
+   2. Unnecessary pinning/unpinning of pages and memory copy/delete can be avoided. If SARGs were checked by the routine that calls access methods, the method needs to either pin a page in the buffer pool or copy a tuple into memory in order to return a handle to the calling routine, an activity which might be a waste if the tuple doesn't satisfy the SARG. Therefore, it is better if the SARG is checked by the access method each page at a time so that only tuples that satisfy the SARG can be returned.
+- Rows in base tables can be referenced by using their physical address on the disk or the row primary key. Using physical address is problematic in case an update requires the row to be moved to a different page. On the other hand, referencing by primary key is slower overall. Both physial address and primary key can be stored to be used in different cases.
+
+### Data Warehouses
```
Comment (twitu) [inline week4/db-architecture-4.md]: For future reference, a number indicating the section of the this topic in the paper would be good.

**Pull request closed without merging**